src/platform/windows/mod.rs

```diff
@@ -0,0 +1,1482 @@
+        self.reader.borrow_mut().cancel_io();
+    }
+}
+
+impl OsIpcReceiver {
+    // We can't just use an anonymous pipe via CreatePipe(), even though
+    // that creation is faster -- we a pipe created with FILE_FLAG_OVERLAPPED.
+    fn new() -> Result<OsIpcReceiver,WinError> {
+        let mut r = try!(OsIpcReceiver::new_named());
+        let pipe_id = mem::replace(&mut r.pipe_id, None).unwrap();
+        r.sender = Some(try!(OsIpcSender::connect_pipe_id(pipe_id)));
+        Ok(r)
+    }
+
+    fn from_handle(handle: HANDLE) -> OsIpcReceiver {
+        OsIpcReceiver {
+            handle: WinHandle::new(handle),
+            sender: None,
+            iocp: INVALID_HANDLE_VALUE,
+            pipe_id: None,
+            reader: RefCell::new(MessageReader::new(handle)),
+        }
+    }
+
+    fn new_named() -> Result<OsIpcReceiver,WinError> {
+        unsafe {
+            // FIXME make_pipe_id is a big performance cost.  We need
+            // a more efficient way of generating pipe IDs.
+            let pipe_id = make_pipe_id();
+            let pipe_name = make_pipe_name(&pipe_id);
+
+            // create the pipe server
+            let handle =
+                kernel32::CreateNamedPipeA(pipe_name.as_ptr(),
+                                           winapi::PIPE_ACCESS_INBOUND | winapi::FILE_FLAG_OVERLAPPED,
+                                           winapi::PIPE_TYPE_BYTE | winapi::PIPE_READMODE_BYTE | winapi::PIPE_REJECT_REMOTE_CLIENTS,
+                                           // 1 max instance of this pipe
+                                           1,
+                                           // out/in buffer sizes
+                                           0, WRITE_BUFFER_SIZE as u32,
+                                           0, // default timeout for WaitNamedPipe (0 == 50ms as default)
+                                           ptr::null_mut());
+            if handle == INVALID_HANDLE_VALUE {
+                return Err(WinError::last("CreateNamedPipeA"));
+            }
+
+            Ok(OsIpcReceiver {
+                handle: WinHandle::new(handle),
+                sender: None,
+                iocp: INVALID_HANDLE_VALUE,
+                pipe_id: Some(pipe_id),
+                reader: RefCell::new(MessageReader::new(handle)),
+            })
+        }
+    }
+
+    fn sender(&mut self) -> Result<OsIpcSender,WinError> {
+        Ok(mem::replace(&mut self.sender, None).unwrap())
+    }
+
+    fn is_closed(&self) -> bool {
+        self.reader.borrow().closed
+    }
+
+    fn prepare_for_transfer(&mut self) -> Result<bool,WinError> {
+        let mut reader = self.reader.borrow_mut();
+        // cancel any outstanding IO request
+        reader.cancel_io();
+        // this is only okay if we have nothing in the read buf
+        Ok(reader.read_buf.is_empty())
+    }
+
+    pub fn consume(&self) -> OsIpcReceiver {
+        let handle = dup_handle(*self.handle).unwrap();
```

Comment:
Correction: the `WinHandle` type itself of course doesn't have interior mutability; but you could wrap it in a `RefCell<>` -- which is what the `mpsc` and `macos` implementations are doing, and what I'll most likely change the `unix` implementation to do as well.

In fact you already *have* one instance of the handle inside a `RefCell`: the one inside `reader`! I have been suspecting for a while -- and am getting more and more positive -- that it might indeed be best to use that one for everything (after turning it into a proper `WinHandle`); and not keep the instance at the top level of `OsIpcReceiver` at all?...